chore: add commit hooks

boneskull opened the pull request:
This change: 

- adds a pre-commit hook that runs `eslint --fix` on any JS file on the stage.
- adds a commit message check that verifies the commit message is in conventional commits format.
- adds a new CI workflow to check commit messages in pull requests _which target `main`_

We will definitely want this if we want better automation of our releases.  Please see [these docs](https://github.com/conventional-changelog/commitlint/#what-is-commitlint) for more info about [commitlint](https://npm.im/commitlint) and how we can further configure it.  I have created a [configuration](https://github.com/LavaMoat/LavaMoat/blob/boneskull/husky/commitlint.config.js) which adopts the [conventional commits](https;//conventionalcommits.org) format but also relaxes some of the more pedantic default rules.

> Note: I'm unsure if using a `prepare` script is OK here.  This command tells `husky` to install the hooks into the `.git/` directory, and is only run in a development environment upon `npm install`

naugtur commented on package.json:
```diff
@@ -22,12 +26,20 @@
+    "lint:staged": "lint-staged",
+    "lint:commit": "commitlint",
     "preversion": "npm run rebuild",
+    "prepare": "husky install",
```
`prepare`  runs on publishing too. Why not `postinstall`?

sidenote: husky is fairly complex. Do we need more than https://www.npmjs.com/package/pre-commit ?
(edit)
Oh, wait - pre-commit won't handle checking the message. nevermind

legobeat replied:
I went for https://github.com/toplenboren/simple-git-hooks last time I had to pick one FWIW. `pre-commit` drags along a bit too much deps too be worth it imo.

naugtur replied:
sounds good. husky is also fine.

Let's address the lifecycle script choice though.

boneskull replied:
~~I think you may be right about `postinstall` if we want to avoid `husky install` on publish--but only because we're in a workspace root.  `prepare` should probably be calling `npm run rebuild` instead of `preversion`, then.~~

wait, because we have `allow-scripts` configured, it doesn't run automatically anyway.

anyhow, I've changed it to this:

- `postinstall` runs `husky install`
- `prepare` runs `rebuild`, which calls the `rebuild` script in workspaces where it is extant
- `preversion` removed
- add `$root = true` to `allowScripts` config

boneskull committed:
chore: add commit hooks and prepare script

- adds a pre-commit hook that runs `eslint --fix` on any JS file on the stage
- adds a commit message check that verifies the commit message is in conventional commits format
- adds `husky install` to workspace root `postinstall` script, to be run via `allow-scripts`
- adds a `prepare` lifecycle script which runs the `rebuild` script
- remove `preversion` script

boneskull committed:
chore(ci): add commitlint workflow

naugtur commented on .github/workflows/commitlint.yml:
```diff
@@ -0,0 +1,17 @@
+name: Validate Commit Messages
+on:
+  [pull_request]
```
+1 to running it as early as possible, not only for things targeting master. 
This may occasionally be annoying when doing messy experiments, but it's totally worth it.

boneskull replied:
lol, I just reverted this.  I'm not sure it is worth it, honestly.  but I'll put it back.  

note that it does not run on `push` because, well, if you've pushed to `main` it's too late anyway.  maybe we need a branch protection rule (if we don't have one already) that disallows direct pushes to `main`.

naugtur reviewed (approved):
Please add a section to the README documenting the commit convention matching commitlint.config.js
